webhook/stats_reporter.go

```diff
@@ -65,43 +65,111 @@ var (
+			return req.Namespace
+		},
+		admissionAllowedKey: func(_ *admissionv1.AdmissionRequest, resp *admissionv1.AdmissionResponse) string {
+			return strconv.FormatBool(resp.Allowed)
+		},
+	}
+	conversionTags = map[tag.Key]conversionToValue{
+		desiredAPIVersionKey: func(req *apixv1.ConversionRequest, _ *apixv1.ConversionResponse) string {
+			return req.DesiredAPIVersion
+		},
+		resultStatusKey: func(_ *apixv1.ConversionRequest, resp *apixv1.ConversionResponse) string {
+			return resp.Result.Status
+		},
+		resultReasonKey: func(_ *apixv1.ConversionRequest, resp *apixv1.ConversionResponse) string {
+			return string(resp.Result.Reason)
+		},
+		resultCodeKey: func(_ *apixv1.ConversionRequest, resp *apixv1.ConversionResponse) string {
+			return strconv.Itoa(int(resp.Result.Code))
+		},
+	}
+)
+
 // StatsReporter reports webhook metrics
 type StatsReporter interface {
 	ReportAdmissionRequest(request *admissionv1.AdmissionRequest, response *admissionv1.AdmissionResponse, d time.Duration) error
 	ReportConversionRequest(request *apixv1.ConversionRequest, response *apixv1.ConversionResponse, d time.Duration) error
 }
 
+type options struct {
+	tagsToExclude map[string]struct{}
+}
+
+type Option func(_ *options)
+
+func WithoutTag(tag string) Option {
+	return func(opts *options) {
+		if opts.tagsToExclude == nil {
+			opts.tagsToExclude = make(map[string]struct{})
+		}
+		opts.tagsToExclude[tag] = struct{}{}
+	}
+}
+
 // reporter implements StatsReporter interface
 type reporter struct {
-	ctx context.Context
+	ctx  context.Context
+	opts options
 }
 
 // NewStatsReporter creates a reporter for webhook metrics
-func NewStatsReporter() (StatsReporter, error) {
+func NewStatsReporter(opts ...Option) (StatsReporter, error) {
 	ctx, err := tag.New(
 		context.Background(),
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	return &reporter{ctx: ctx}, nil
+	options := options{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &reporter{ctx: ctx, opts: options}, nil
 }
 
 // Captures req count metric, recording the count and the duration
 func (r *reporter) ReportAdmissionRequest(req *admissionv1.AdmissionRequest, resp *admissionv1.AdmissionResponse, d time.Duration) error {
-	ctx, err := tag.New(
-		r.ctx,
-		tag.Insert(requestOperationKey, string(req.Operation)),
-		tag.Insert(kindGroupKey, req.Kind.Group),
-		tag.Insert(kindVersionKey, req.Kind.Version),
-		tag.Insert(kindKindKey, req.Kind.Kind),
-		tag.Insert(resourceGroupKey, req.Resource.Group),
-		tag.Insert(resourceVersionKey, req.Resource.Version),
-		tag.Insert(resourceResourceKey, req.Resource.Resource),
-		tag.Insert(resourceNamespaceKey, req.Namespace),
-		tag.Insert(admissionAllowedKey, strconv.FormatBool(resp.Allowed)),
-	)
+	mutators := []tag.Mutator{}
+
+	for key, f := range admissionTags {
+		if _, ok := r.opts.tagsToExclude[key.Name()]; ok {
+			continue
+		}
+		mutators = append(mutators, tag.Insert(key, f(req, resp)))
+	}
```

Comment:
Can we compute this list once so we're not doing this on every request but instead it's computed when the `StatsReporter` is initialized.

Reply:
If I understand this correctly, this can't be pre-computed, as this is where tag values are actually computed (per request).

Reply:
Make sense! I moved this computation into `NewStatsReporter`.